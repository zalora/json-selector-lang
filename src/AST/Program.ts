// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTNode from './ASTNode';

// Program is the root node of the abstract syntax tree representing the JSL input.
class Program implements ASTNode {
  statements: Array<ASTNode>;

  constructor() {
    this.statements = [];
  }

  tokenLiteral(): string {
    if (this.statements.length <= 0) {
      return '';
    }

    return this.statements[0]?.tokenLiteral() || '';
  }

  toString(): string {
    return this.statements.reduce((prev, next) => prev + `${next.toString()}`, '');
  }
}

export default Program;
