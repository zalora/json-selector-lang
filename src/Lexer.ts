// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import { dot, int, ident, lbracket, rbracket, eof } from './Tokens';

interface Token {
  type: Symbol;
  literal: string;
}

// Lexer represents the lexical analyzer for JSON Selector Language
class Lexer {
  private input: Array<string> = [];
  private position: number = 0;
  private readPosition: number = 0;
  private currentCharacter: string = '';

  constructor(input: string) {
    this.input = Array.from(input);
    this.readChar();
  }

  //  nextToken returns the next lexed token for the given input
  public nextToken(): Token {
    const ch = this.currentCharacter;
    if (!ch) {
      return { type: eof, literal: '' };
    }

    let token: Token;

    switch (ch) {
      case '.':
        token = { type: dot, literal: ch };
        break;
      case '[':
        token = { type: lbracket, literal: ch };
        break;
      case ']':
        token = { type: rbracket, literal: ch };
        break;
      default:
        if (this.isNumeric(ch)) {
          const num = this.readNumber();
          return { type: int, literal: num };
        }
        return this.readIdentifier();
    }

    this.readChar();

    return token;
  }

  /**
   * readChar reads one character
   * Note that only ASCII is supported.
   */
  private readChar() {
    if (this.readPosition >= this.input.length) {
      this.currentCharacter = '';
    } else {
      this.currentCharacter = this.input[this.readPosition];
    }

    this.position = this.readPosition;
    this.readPosition += 1;
  }

  // isNumeric helps identify if the given string is a number or not
  private isNumeric = (n: string): boolean => {
    const num: number = parseFloat(n);

    return !isNaN(num) && isFinite(num);
  };

  // readIdentifier returns a token with an identifier as the literal
  private readIdentifier(): Token {
    let startIdx = this.position;
    while (this.isIdent(this.currentCharacter)) {
      this.readChar();
    }

    const identfier: Array<string> = this.input.slice(startIdx, this.position);
    const identStr: string = identfier.join('');

    return { type: ident, literal: identStr };
  }

  // readNumber returns a token with a numeric literal
  private readNumber(): string {
    let startIdx = this.position;
    while (this.isNumeric(this.currentCharacter)) {
      this.readChar();
    }

    const number: Array<string> = this.input.slice(startIdx, this.position);

    return number.join('');
  }

  // isIdent returns true if a character is not one of the other tokens
  private isIdent = (ch: string | null): boolean => {
    if (!ch) {
      return false;
    }

    return ch !== '.' && ch !== '[' && ch !== ']';
  };
}

export default Lexer;
