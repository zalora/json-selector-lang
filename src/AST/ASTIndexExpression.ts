// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTIndexExpression implements ASTExpression {
  token: Token;
  left: ASTExpression;
  index: ASTExpression;

  constructor(token: Token, left: ASTExpression, index: ASTExpression) {
    this.token = token;
    this.left = left;
    this.index = index;
  }

  getDescription(): string {
    const { description: leftDesc = '' } = this.left || {};
    const { description: indexDesc = '' } = this.index || {};

    if (!leftDesc || !indexDesc) {
      return '';
    }

    return `(${this.left.description}[${this.index.description}])`;
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTIndexExpression;
