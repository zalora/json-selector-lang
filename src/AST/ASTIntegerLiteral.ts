// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTIntegerLiteral implements ASTExpression {
  token: Token;
  value: string;

  constructor(token: Token, value: string) {
    this.token = token;
    this.value = value;
  }

  getDescription(): string {
    return this.value || '';
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTIntegerLiteral;
