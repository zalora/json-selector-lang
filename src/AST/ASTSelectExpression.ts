// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import ASTExpression from './ASTExpression';
import { Token } from '../Lexer';

class ASTSelectExpression implements ASTExpression {
  token: Token;
  key: string;

  constructor(token: Token, key: string) {
    this.token = token;
    this.key = key;
  }

  getDescription(): string {
    return `${this.token.literal} ${this.key}`;
  }

  tokenLiteral(): string {
    const { literal = '' } = this.token || {};

    return literal;
  }
}

export default ASTSelectExpression;
