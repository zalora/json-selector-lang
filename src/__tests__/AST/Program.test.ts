// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import Lexer from '../../Lexer';
import Program from '../../AST/Program';
import ASTSelectExpression from '../../AST/ASTSelectExpression';
import ASTIntegerLiteral from '../../AST/ASTIntegerLiteral';
import ASTIndexExpression from '../../AST/ASTIndexExpression';

describe('Program', () => {
  const inputSelect = '.data';
  const lSelect = new Lexer(inputSelect);
  const pSelect = new Program();
  const tokenFirst = lSelect.nextToken();
  const tokenSecond = lSelect.nextToken();
  const selectExpr = new ASTSelectExpression(tokenFirst, tokenSecond.literal);
  pSelect.statements.push(selectExpr);

  it('tests with token input as select expression', () => {
    expect(pSelect).toMatchSnapshot();
  });

  it('tests if tokenLiteral returns literal for token input as select expression', () => {
    expect(pSelect.tokenLiteral()).toMatchSnapshot();
  });

  it('tests if toString returns all the list of statements for token input as select expression', () => {
    expect(pSelect.toString()).toMatchSnapshot();
  });

  const inputIndex = '.images[2]';
  const lIndex = new Lexer(inputIndex);
  const pIndex = new Program();
  const tokenDot = lIndex.nextToken();
  const tokenIdent = lIndex.nextToken();
  const tokenLbracket = lIndex.nextToken();
  const tokenInt = lIndex.nextToken();
  const lExpr = new ASTSelectExpression(tokenDot, tokenIdent.literal);
  const rExpr = new ASTIntegerLiteral(tokenInt, 2);
  const indexExpr = new ASTIndexExpression(tokenLbracket, lExpr, rExpr);
  pIndex.statements.push(indexExpr);

  it('tests with token input as index expression', () => {
    expect(pIndex).toMatchSnapshot();
  });

  it('tests if tokenLiteral returns literal for token input as index expression', () => {
    expect(pIndex.tokenLiteral()).toMatchSnapshot();
  });

  it('tests if toString returns all the list of statements for token input as index expression', () => {
    expect(pIndex.toString()).toMatchSnapshot();
  });
});
