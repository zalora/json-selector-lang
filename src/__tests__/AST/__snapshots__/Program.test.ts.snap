// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Program tests if toString returns all the list of statements for token input as index expression 1`] = `"(.images[2])"`;

exports[`Program tests if toString returns all the list of statements for token input as select expression 1`] = `".data"`;

exports[`Program tests if tokenLiteral returns literal for token input as index expression 1`] = `"["`;

exports[`Program tests if tokenLiteral returns literal for token input as select expression 1`] = `"."`;

exports[`Program tests with token input as index expression 1`] = `
Program {
  "statements": Array [
    ASTIndexExpression {
      "index": ASTIntegerLiteral {
        "token": Object {
          "literal": "2",
          "type": Symbol(int),
        },
        "value": 2,
      },
      "left": ASTSelectExpression {
        "key": "images",
        "token": Object {
          "literal": ".",
          "type": Symbol(.),
        },
      },
      "token": Object {
        "literal": "[",
        "type": Symbol([),
      },
    },
  ],
}
`;

exports[`Program tests with token input as select expression 1`] = `
Program {
  "statements": Array [
    ASTSelectExpression {
      "key": "data",
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
  ],
}
`;
